### ISSUE TITLE
Formatting guidelines for inline assembly

### ISSUE THREAD

**Issue by joshtriplett** [rust-dev-tools/fmt-rfcs#152]
An issue I've observed in several contexts discussing the new `asm!` syntax for inline assembly: everyone formats `asm!` statements differently, and we should 1) come up with guidance on how to do so, and 2) implement that guidance in `rustfmt`.

Notably, this includes how to format both single-line and multi-line assembly statements.

With my style team hat on, I would propose the following guidelines:
- Single-line assembly strings should be treated as any other argument to a format macro.
- Multi-line assembly strings should be formatted with the opening quote right after the paren like `asm!("`, and the closing quote indented to match the arguments:
    ```rust
    asm!("
        instruction
        instruction
        ",
        ...
    );
    ```
- Simple one-instruction `asm!` blocks can have the arguments on the same line, if the whole thing from `asm!(` to `);` (plus indentation) fits in the line width.
- Otherwise, one argument per line with trailing commas, indented one level past `asm!`, just like a function.
- `options(...)` goes on one line if it fits, and otherwise format it as though it were a nested function call to a function named `options`.
- Never place any space or line breaks inside of `in(reg)` or `out(reg)` or `inout(reg)` or `in("regname")` or `out("regname")` or similar; always treat them as a single atomic unit.
- If an `inout` or `inlateout` pair of expressions are too long to fit on one line, break before (never after) the `=>`, and indent the `=>` one level further:
    ```rust
    asm!(
        "instruction {}",
        inout(reg) very_long_expression
            => very_long_out_expression,
    )
    ```
- If an `in(reg)` or `out(reg)` or `lateout(reg)` expression is too long, break between the `)` and the expression and indent one level, then format from there; however, if the expression can be broken internally, follow same the rules for when to leave the head of the expression on the same line as the `in(reg)` or similar as if `in(reg)` were the opener of a function:
    ```rust
    asm!(
        "instruction {}",
        in(reg)
            extremely_long_unbreakable_expression,
        in(reg) function_name()
            .method_name(method_arg)
            .further_chained_method(),
        out(reg) long_function_name(
            long_function_argument_expression,
        ),
    );
    ```
- For named arguments like `name = in(reg) expression`, line-break it as you would an assignment statement with the same indentation, treating the `in(reg) expression` as the right-hand side of the assignment, and following all the same rules as above.

If we were to, in the future, introduce a syntax for format macros that allows putting a set of arguments (if that fit) on the same line as the `",` ending the string, we should do the same for `asm!`. We shouldn't introduce special formatting for one without the other, though.

**Comment by Amanieu** [comment]
Formatting the string literal is actually very tricking since we technically need to preserve whitespace and newlines when formatting. This will limit how much we can we format the template string, in particular regarding indentation.

We could make an exception here since whitespace isn't significant for inline assembly.

**Comment by joshtriplett** [comment]
@Amanieu That's not *entirely* true, if you do something especially unusual in your inline assembly. Assembly supports multi-line `.string` directives, among likely other things:

```rust
#![feature(asm)]

fn main() {
    let mut i: u32 = 2;
    unsafe {
        asm!(
            r#"
            add {:e}, 3
            jmp 1f
            .string "hello
            world"
            1:
            "#,
            inout(reg) i,
        );
    }
    dbg!(i);
}
```

(Assume for the sake of argument that I used that string for something.)

If you look at the disassembly of this, the whitespace is significant. Changing the amount of whitespace at the start of lines would change the semantic meaning of the program.

One way around this would be if we recommended using `concat!` for multi-line assembly. I don't know that that's worth the tradeoff.

Initially I would propose having formatting guidelines, and then doing as much as we can in `rustfmt` without changing the string literal.

I can absolutely think of real-world reasons someone might write something like this; for instance, embedding some long buffer. I can also think of better ways to do that, but nonetheless, formatting must not break user code. So I don't think we can consider any formatting that changes even the amount of space inside the string. In theory, we *might* be able to get away with changing the amount of whitespace at the start and end of the string, *maybe*, but I'm not sure that's worth it, and we'd still have to think *very* hard to see if anyone could write asm code that that would break.

**Comment by Amanieu** [comment]
I think that the case for formatting multi-line strings is more general and not restricted to inline asm: https://github.com/rust-lang/rustfmt/issues/2876

**Comment by BartMassey** [comment]
I think I would like to see something along the lines of the weird C "adjacent string constants" thing for this case. Then I could put one string per line, which seems to me more convenient. Alternatively, we could just make the macro take an arbitrary number of strings, but I fear much confusion at the end there.

Maybe go nuts and have an implicit newline when the last character is not whitespace and an implicit tab when the first character is not whitespace? The whole

```
"
xor %rax, %rax\n
\tmov %rax, %rcx\n
.l0:\n
\tinc %rcx\n
"
```

nonsense I tend to go through with GCC/Clang is pretty annoying.

Speaking of which, what *is* the assumed whitespace state just before and just after a new `asm!()` statement?

**Comment by joshtriplett** [comment]
If you want separate strings for each line, you can do that with `concat!`. That would certainly make formatting easier, at a cost of verbosity and extra indentation.

The assumed whitespace state should indeed be documented. There will be a newline before and after the assembly.

**Comment by Amanieu** [comment]
I don't think that whitespace state before/after the asm is significant since whitespace (except for newlines) is ignored by all assembly languages.

The use of "\t" in C is just because GCC pastes inline assembly directly into the output .S file and some developers want the -S output of the compiler to be nicely aligned when reading the resulting assembly. This does not apply to LLVM since it parses the inline asm and then re-prints the instructions from scratch when generating assembly output.

**Comment by BartMassey** [comment]
```rust
#![feature(asm)]

macro_rules! asm_block {
    ({$($line:literal),*$(,)?}$($stuff:tt)*) => {
        asm!(concat!($(concat!($line,"\n")),*)$($stuff)*)
    };
}

fn main() {
    let x: u64;
    unsafe { asm_block!(
        {
            "nop",
            "xor {0},{0}",
        },
        out(reg) x,
    )};
    println!("{}", x);
}
```

I will admit to being one of those people who would like to see lines tabbed by default in case I want to read the assembly. The macro could be fancied to do this either by providing an explicit no-tab or by inferring: e.g. lines that begin with "." or end with ":" don't get tabbed.

**Comment by Amanieu** [comment]
I *really* like this syntax and I'm considering supporting it directly in `asm!`. I would just make a small modification: adjacent literals concatenate directly and `;` concatenates with a newline:

```rust
fn main() {
    let x: u64;
    unsafe { 
        asm!(
            {
                "nop";
                "xor" " {0}," " {0}";
            },
            out(reg) x,
        );
    }
    println!("{}", x);
}
```

I've used `concat!` a lot in the past with `llvm_asm!` but in my experience this has been rather troublesome:

```rust
macro_rules! asm_read {
    ($instr:expr, $width:expr, $trapped:expr, $type:ty, $ptr:expr) => {{
        let tmp: $type;
        llvm_asm! {
            concat!(
                "0: ", $instr, " ${1:", $width, "}, [$2];",
                asm_trap_list!()
            )
            : "+{x16}" ($trapped), "=r" (tmp)
            : "r" ($ptr as u64)
            :
            : "volatile"
        };
        mem::transmute_copy(&tmp)
    }};
}
```

With the new style it would look like this:

```rust
macro_rules! asm_read {
    ($instr:expr, $width:expr, $trapped:expr, $type:ty, $ptr:expr) => {{
        let tmp: $type;
        asm!(
            {
                "0: " $instr " {out:" $width "}, [{ptr}]";
                asm_trap_list!();
            },
            out = lateout(reg) tmp,
            ptr = in(reg) $ptr,
            inout("x16") $trapped,
            options(nostack, preserves_flags),
        );
        mem::transmute_copy(&tmp)
    }};
}
```

**Comment by BartMassey** [comment]
If you want to allow multiple literals per line, I'd suggest adding a space between each of them internally — I think it's going to be less error-prone.

**Comment by joshtriplett** [comment]
@BartMassey 
> I will admit to being one of those people who would like to see lines tabbed by default in case I want to read the assembly.

If you mean the assembly output from the compiler, it sounds like it will get properly formatted (by LLVM) even if the input is not.

@Amanieu I really like the idea of a syntax like this as well. I would suggest a slight tweak, though: rather than putting a braced block as the first argument, could we just turn the whole `asm!` into a braced block rather than a parenthesized one?

```rust
asm!{
    "instruction 1",
    "instruction 2",
    "instruction 3",
    out = lateout(reg) var,
    options(nostack),
}
```

(I don't want to bikeshed the separators between assembly strings, though I will admit that I find a mix of semicolons and commas not ideal)

**Comment by BartMassey** [comment]
I guess I could have written the macro to avoid the braced block; I was just being lazy when I put it in, but since those strings are the only arguments of token type `literal` I guess I didn't need the braces. Ah well.

**Comment by Amanieu** [comment]
Note that macros don't care about the type of delimiter used: `[]`, `()` and `{}` can all be used (this is how `vec![]` works).

I like the idea about allowing multiple string literals in `asm!` which would be treated as separate lines. I feel that this would make multi-line asm much more readable than multi-line strings.

**Comment by joshtriplett** [comment]
@Amanieu It'll also mean that rustfmt will be able to reliably format inline asm blocks. I'd love to see this.

**Comment by joshtriplett** [comment]
Implemented in https://github.com/rust-lang/rust/pull/73364 ; RFC updated via https://github.com/Amanieu/rfcs/pull/1 .

**Comment by joshtriplett** [comment]
I've updated the formatting guidelines at the top of this issue to use the new support for multiple assembly string arguments.

**Comment by calebcartwright** [comment]
@joshtriplett - is my understanding correct that you all have settled on target formatting as summarized in the updated description, and all that's pending is for someone to codify those rules within the guide? (wondering if we need a new page for special case macros that have their own spec :thinking:)

**Comment by joshtriplett** [comment]
I believe we've settled on exactly what's in the description of this issue, yes.

**Comment by calebcartwright** [comment]
cc @ytmimi - know we've got a few things already in flight, but think this would be a good item for you to look at next if you're interested. we can chat more about it offline

**Comment by ytmimi** [comment]
Hey @joshtriplett, I've spent some time working on getting this implemented in rustfmt (rust-lang/rustfmt#5191). The guidelines you wrote up were a great help. I also found the [asm section of the Unstable Book](https://doc.rust-lang.org/unstable-book/library-features/asm.html) useful!

Templates and options were well specified, but I have a few clarifying questions around the ``operands`` and ``clobber_abis`` that I hope you can help me with.

#### Operands

How should we handle ``const`` and ``sym`` operands?

Is it better to not break between the keyword (const / sym) and the expression for long, unbreakable expressions?
```rust
asm!(
    "instruction {}",
    const extremely_long_unbreakable_expression,
    sym extremely_long_unbreakable_symbol_name,
)
```

In the case that we're dealing with a named argument should we treat const and sym operands as follows?

```rust
asm!(
    "instruction {}",
    long_named_argument =
        const extremely_long_unbreakable_expression,
    long_named_argument =
        sym extremely_long_unbreakable_symbol_name,
);
```

For const specifically, is this the desired formatting for long expression that can be broken? Given that sym is meant to refer to a function name, we don't need to worry about these cases, right?

```rust
asm!(
    "instruction {}",
    const function_name()
        .method_name(method_arg)
        .further_chained_method(),
);

asm!(
    "instruction {}",
    const long_function_name(
        long_function_argument_expression
    ),
);
```

Here are some other test cases that I wanted to get your opinion on:

```rust
asm!(
    "instruction {}",

    // case 1 - in chain out expr
    inout(reg) function_name()
        .method_name(method_arg)
        .further_chained_method()
        => very_long_out_expression,

    // case 2 - in function our expr
     inout(reg) long_function_name(
        long_function_argument_expression
    )
        => very_long_out_expression,

    // case 3 - in expr out chain
    inout(reg) very_long_expression
        => function_name()
            .method_name(method_arg)
            .further_chained_method(),

    // case 4 - in chain out chain
    inout(reg) function_name()
        .method_name(method_arg)
        .further_chained_method()
        => function_name()
            .method_name(method_arg)
            .further_chained_method(),

    // case 5 - in function out chain
    inout(reg) long_function_name(
        long_function_argument_expression
    )
        => function_name()
            .method_name(method_arg)
            .further_chained_method(),

    // case 6 - in expr out function
    inout(reg) very_long_expression
        => long_function_name(
            long_function_argument_expression
        ),

    // case 7 - in chain out function
    inout(reg) function_name()
        .method_name(method_arg)
        .further_chained_method()
        => long_function_name(
            long_function_argument_expression
        ),

    // case 8 - in function out function
    inout(reg) long_function_name(
        long_function_argument_expression
    )
        => long_function_name(
            long_function_argument_expression
        ),
)
```

For completeness I added test cases for ``inout`` and ``inlateout`` that also include named arguments. Just want to double check that something like this is okay.

```rust
asm!(
inout(reg) long_function_name(
    aaaaaaaaaaaaaaaaaaaa =
        inout(reg) very_long_expression
            => very_long_out_expression,
);
```

#### Clobber ABIs

The [Reference-level explanation](https://doc.rust-lang.org/unstable-book/library-features/asm.html#reference-level-explanation) in the Unstable book gives the following ABNF for ``clobber_abis`` and ``options``:

```
clobber_abi := "clobber_abi(" <abi> *["," <abi>] [","] ")"
options := "options(" option *["," option] [","] ")"
```

Given that the ABNF is so similar should we also treat clobber_abis as we do options? So treat them like nested function calls if we need to break them.


Lastly, how should we handle trailing commas? I know that for most macros we don't want to add or remove tokens, but in this case trailing commas are optional. Is there a recommendation for how we should handle trailing commas?

**Comment by ytmimi** [comment]
A few more operand related test cases that I wanted to get your opinion on:

In this case we have an ``inout`` that doesn't fit on the same line, but the ``out_expr`` is ``_``. Should we still force an indented newline after the in expression?
```rust
asm!(
    "instruction {}",
    inout(reg) long_function_name(
        long_function_argument_expression
    )
        => _,
);
```
Or does this look better:
```rust
asm!(
    "instruction {}",
    inout(reg) long_function_name(
        long_function_argument_expression
    ) => _,
);
```

Same question as above, but with chains:

```rust
asm!(
    "instruction {}",
    in(reg) function_name()
        .method_name(method_arg)
        .further_chained_method()
        => _,
);
```

```rust
asm!(
    "instruction {}",
    in(reg) function_name()
        .method_name(method_arg)
        .further_chained_method() => _,
);
```

Do we always want to break inout's if the in_expr and out_expr don't fit on the same line? Here's an example where the two technically don't fit but I'm not sure if it's better to force the break or not:
```rust
asm!(
    "instruction {}",
    inout(reg) very_long_expression => function_name(
        long_function_argument_expression
    ),
);
```
Or do we always force the break?
```rust
asm!(
    "instruction {}",
    inout(reg) very_long_expression
        => function_name(
            long_function_argument_expression
        ),
);
```

Again, thanks for helping to clarify this for me.

**Comment by calebcartwright** [comment]
@ytmimi - I'd suggest we move forward with a PR in this repo that adds text to the guide which covers the above. For the outstanding cases feel free to make your suggested guidance in the respective wording.

Folks can weigh in on that PR, and we can drop notes in a few different channels for broader awareness to allow others to weigh in. We'll also have a little flexibility for a while to adjust things, especially given that we have the benefit in this particular instance of being able to progressively roll out this behavior in rustfmt via a config option.

I think the existing guide text and structure should provide a good reference for framing the resultant rules identified on this thread, but feel free to ping me offline to chat about any specifics

**Comment by ppannuto** [comment]
One critical absence from this discussion and all the examples in the thread so far is any guidance for **comments** explaining the assembly code.

I did a rough cut of [converting current assembly in Tock to the proposed syntax](https://github.com/tock/tock/pull/4515/files), and splitting assembly lines into individual strings makes managing, parsing, and editing comments really challenging.

I'm not quite sure I understand the impetus for splitting one conceptual block of assembly code into many separate strings—are the some examples of larger assembly blocks where separate strings improves things? I think all the short examples here hide how painful that is in practice.

@joshtriplett I'm curious if you could take a look at some of the code in the linked diff here and weigh on the desired style for comments that support larger and more complex assembly. Assembly is famously hard to get right at baseline, and I think any recommended style needs to consider explanatory comments and ergonomic formatting and editing as a first-class concern.

Maybe an easy fix would be guidance to "right-pad" the block with trailing whitespace so that all the lines end up the same length? The intermingled and semantically meaningless (for the assembly you are reading the comment to understand at least) `",` throughout comment blocks is a big part of what makes reading this challenging.

**Comment by programmerjake** [comment]
we could put comments outside the strings, maybe like so (adapted from the [PR](https://github.com/tock/tock/pull/4515/files) @ppannuto wrote):
```rust
        "100:", // bss_init_loop
        // We increment r0. Check if we have reached r1 (end of .bss), and stop if so.
        "  cmp r1, r0",
        "  beq 101f", // If r0 == r1, we are done.
        // Write a word to the address in r0, and increment r0.
        // Since r2 contains zero, this will clear the memory pointed to by r0.
        // Using `stm` (store multiple) with the bang allows us to also increment r0 automatically.
        "  stm r0!, {{r2}}",
        "  b 100b", // Continue the loop.
```

**Comment by BartMassey** [comment]
@programmerjake Yes, this is the comment style I normally use. It is nice to be able to use standard Rust editors for managing assembly comments.

@ppannuto I personally prefer to write longer assembly code in a separate file and `include_str!()` it, or assemble it in `build.rs`.

The big advantage of one string per line is that all of Rust's editing and formatting machinery has access to it: changing indentation, `rustfmt`, etc "just works" for the most part. This is great when the primary language in a source file is Rust and the assembly is just sprinkled in, which is the normal case for inline asm. When the primary language is assembly, I find it nicer to just write some assembly using my editor's tools for that, assembling as I go to see if the syntax checks out.

I personally don't really struggle with "reading past" the string quotes. That said, there's a long tradition of whitespace and other padding character adventures in assembly code: I wouldn't be opposed to allowing arbitrary left and right space padding to uniform length (although I personally would prefer not to do that).
